src/darwin/Framework/CHIP/MTRAsyncCallbackWorkQueue.mm

```diff
@@ -169,9 +169,51 @@ - (void)_callNextReadyWorkItem
         self.runningWorkItemCount = 1;
 
         MTRAsyncCallbackQueueWorkItem * workItem = self.items.firstObject;
+
+        // Check if batching is possible or needed. Only ask work item to batch once for simplicity
+        if (workItem.batchable && workItem.batchingHandler && (workItem.retryCount == 0)) {
+            while (self.items.count >= 2) {
+                MTRAsyncCallbackQueueWorkItem * nextWorkItem = self.items[1];
+                if (!nextWorkItem.batchable || (nextWorkItem.batchingID != workItem.batchingID)) {
+                    // next item is not eligible to merge with this one
+                    break;
+                }
+
+                BOOL fullyMerged = NO;
+                workItem.batchingHandler(workItem.batchableData, nextWorkItem.batchableData, &fullyMerged);
+                MTR_LOG_DEFAULT(
+                    "MTRAsyncCallbackWorkQueue: merged items for batching - fully merged %@", fullyMerged ? @"YES" : @"NO");
+                if (!fullyMerged) {
+                    // if some parts of the next item is
+                    break;
+                }
+
+                [self.items removeObjectAtIndex:1];
+            }
+        }
+
         [workItem callReadyHandlerWithContext:self.context];
     }
 }
+
+- (BOOL)isDuplicateForTypeID:(NSUInteger)opaqueDuplicateTypeID workItemData:(id)opaqueWorkItemData
+{
+    os_unfair_lock_lock(&_lock);
+    int i = 0;
+    for (MTRAsyncCallbackQueueWorkItem * item in self.items) {
```

Comment:
So merging can only merge in the very next thing, but duplicate detection checks all the things?

Concretely: if the API client does, in this order:

1) Read attribute X
2) Write attribute X
3) Read attribute X

then this duplicate detection will drop the second read, no?  That does not seem right....

Also: self.items[0] might already be running here, right?  Dropping things that match it might be a bit odd in that case... except for reads it might in fact be OK, I guess.

Reply:
Okay this is a real concern. Let me fix the rest of the other points and then think about this.

Reply:
The change to add `BOOL *stop` to the duplicate check handler and have the isDuplicate method start from the end of item array should handle this.